File: autoPublish.py
from logging import log
import random
import time
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options


def createHeadLessDriver():
    """初始化无头浏览器"""
    chrome_options = Options()
    # chrome_options.add_argument("--headless")
    chrome_options.add_argument("--no-sandbox")
    chrome_options.add_argument("--disable-dev-shm-usage")
    chrome_options.add_argument("--disable-gpu")
    chrome_options.add_argument("--window-size=1920,1080")
    
    # 如果需要代理
    # chrome_options.add_argument("--proxy-server=http://your-proxy:port")
    chrome_prefs = {
        "profile.default_content_setting_values.geolocation": 2  # 2 表示拒绝， 1 允许
    }
    chrome_options.add_experimental_option("prefs", chrome_prefs)
    

    return webdriver.Chrome(options=chrome_options)

def createChromeDriver():
    chrome_options = Options()
   
    chrome_prefs = {
        "profile.default_content_setting_values.geolocation": 2  # 2 表示拒绝， 1 允许
    }
    chrome_options.add_experimental_option("prefs", chrome_prefs)
    
    return webdriver.Chrome(options=chrome_options)

class ToutiaoAutoPublisher:
    def __init__(self, driver=createChromeDriver()):
        self.driver = driver
        self.wait = WebDriverWait(self.driver, 20)
        
        # 重要提示：请勿在代码中明文存储账号信息
        self.username = "17355605329"  # 建议从环境变量读取
        self.password = "Aa123456"  # 建议从环境变量读取

    def login(self):
        """处理登录流程"""
        self.driver.get("https://mp.toutiao.com/")
        
        # 等待并选择登录方式（示例为账号密码登录）
        self.wait.until(EC.presence_of_element_located(
            (By.XPATH, "//li[@aria-label='账密登录']")
        ))
        self.driver.find_element(By.XPATH, "//li[@aria-label='账密登录']").click()

        # self.wait.until(EC.element_to_be_clickable(
        #     (By.Name, "//div[contains(text(),'密码登录')]"))).click()
        
        # 输入凭据
        self.wait.until(EC.presence_of_element_located(
            (By.NAME, "normal-input"))).send_keys(self.username)
        self.wait.until(EC.presence_of_element_located(
            (By.NAME, "button-input"))).send_keys(self.password)
        
       # 勾选协议框 <span class="web-login-confirm-info__checkbox" role="checkbox" aria-checked="false" tabindex="0" aria-label="协议勾选框"></span>
        self.driver.find_element(By.CLASS_NAME, "web-login-confirm-info__checkbox").click() 
       
        # 提交登录（注意验证码处理）
        self.driver.find_element(By.CLASS_NAME, "web-login-button").click()
        
        # 等待登录成功
        self.wait.until(EC.url_contains("profile_v4/index"))

    # def publish_drafts(self):
    #     """处理草稿发布流程"""
    #     self.driver.get("https://mp.toutiao.com/profile_v4/manage/draft")
    #     self.wait.until(EC.presence_of_element_located(
    #         (By.XPATH, "//div[contains(@class, 'list')]")
    #     ))

  
    #     while True:
    #         # 获取所有草稿项
    #         drafts = self.wait.until(EC.presence_of_all_elements_located(
    #             (By.XPATH, "//div[contains(@class, 'draft-item')]")
    #         ))

    #         if not drafts:
    #             log.info("没有更多草稿了")
    #             # sleep 1hour
    #             time.sleep(3600)
    #             continue
            
    #         for draft in drafts:
    #             try:
    #                 # 点击编辑按钮
    #                 edit_button = draft.find_element(By.XPATH, ".//a[contains(@class, 'op-button') and text()='编辑']")
    #                 edit_button.click()

    #                 # # 等待编辑页面加载
    #                 self.wait.until(EC.presence_of_element_located(
    #                     (By.XPATH, "//div[contains(@class, 'publish-box')]")))

    #                 # 点击发布按钮
    #                 publish_button = self.wait.until(EC.element_to_be_clickable(
    #                     (By.CLASS_NAME, "publish-content")))
    #                 publish_button.click()

    #                 log.info(f"{time.strftime('%Y-%m-%d %H:%M:%S')} 发布成功")
                    
    #                 # 随机间隔等待 5 ~ 10 min
    #                 snooze = 150 + random.random() * 400
    #                 log.info(f"等待 {snooze} 秒")
    #                 time.sleep(snooze)

    #             except Exception as e:
    #                 print(f"处理草稿时出错: {str(e)}")
    #                 continue  # 继续处理下一个草稿
    def publish_drafts(self):
        """处理草稿发布流程"""
        self.driver.get("https://mp.toutiao.com/profile_v4/manage/draft")
        self.wait.until(EC.presence_of_element_located(
            (By.XPATH, "//div[contains(@class, 'list')]")
        ))

        while True:
            # 获取所有草稿项
            drafts = self.wait.until(EC.presence_of_all_elements_located(
                (By.XPATH, "//div[contains(@class, 'draft-item')]")
            ))

            if not drafts:
                log.info("没有更多草稿了")
                # sleep 1hour
                time.sleep(3600)
                continue
            
            for draft in drafts:
                try:
                    # 获取当前窗口句柄
                    current_window = self.driver.current_window_handle
                    editing_window = None

                    # 点击编辑按钮
                    edit_button = draft.find_element(By.XPATH, ".//a[contains(@class, 'op-button') and text()='编辑']")
                    edit_button.click()

                  
                    # 等待新标签页加载完成
                    self.wait.until(EC.number_of_windows_to_be(2))  # 确保有两个窗口
                    
                    # 切换到新标签页
                    for handle in self.driver.window_handles:
                        if handle != current_window:
                            self.driver.switch_to.window(handle)
                            edgeing_window = handle
                            break

                    # 等待编辑页面加载
                    self.wait.until(EC.presence_of_element_located(
                        (By.XPATH, "//div[contains(@class, 'publish-box')]")))

                   
                    # 使用XPath选择按钮
                    # 等待并获取按钮
                    publish_button = self.wait.until(EC.presence_of_element_located(
                        (By.XPATH, "//button[contains(@class, 'publish-content')]")
                    ))

                    # 滚动到按钮
                    self.driver.execute_script("arguments[0].scrollIntoView(true);", publish_button)

                    # 等待遮挡元素消失
                    self.wait.until(EC.invisibility_of_element_located(
                        (By.CLASS_NAME, "byte-drawer-scroll")
                    ))

                    # 等待按钮可点击
                    self.wait.until(EC.element_to_be_clickable(
                        (By.XPATH, "//button[contains(@class, 'publish-content')]")
                    ))

                    # 点击按钮
                    publish_button.click()
                    time.sleep(3)
                    
                    # 切换回原窗口（如果需要继续处理草稿）
                    self.driver.close()
                    self.driver.switch_to.window(current_window)

                    # 随机间隔等待 3 ~ 10 min
                    snooze = 150 + random.random() * 400
                    print(f"等待 {snooze} 秒")
                    time.sleep(snooze)

                   


                except Exception as e:
                    print(f"处理草稿时出错: {str(e)}")
                    continue  # 继续处理下一个草稿

    def run(self):
        try:
            self.login()
            self.publish_drafts()
        except Exception as e:
            print(f"执行出错: {str(e)}")
        finally:
            self.driver.quit()

if __name__ == "__main__":
    publisher = ToutiaoAutoPublisher(createHeadLessDriver())
    publisher.run()

File: autoPublish.go
package main

import (
	"context"
	"log"
	"strings"
	"time"

	cdp "github.com/chromedp/cdproto/cdp"
	target "github.com/chromedp/cdproto/target"
	"github.com/chromedp/chromedp"
)

type DraftPublisher struct {
	publishController PublisherController
}

func createHeadlessDriver() (context.Context, context.CancelFunc) {
	// 设置 Chrome 的选项（无头模式）
	opts := append(chromedp.DefaultExecAllocatorOptions[:],
		chromedp.Flag("headless", false), // 可选：是否无头模式
		chromedp.Flag("no-sandbox", true),
		chromedp.Flag("disable-dev-shm-usage", true),
		chromedp.Flag("disable-gpu", true),
		chromedp.Flag("window-size", "1920,1080"),
		chromedp.Flag("disable-geolocation", true), // 禁用地理位置
	)

	// 创建一个新的 Chrome 上下文和取消函数
	allocator, cancel := chromedp.NewExecAllocator(context.Background(), opts...)
	ctx, _ := chromedp.NewContext(allocator)

	return ctx, cancel
}

func waitLogin(ctx context.Context) error {
	var username = "17355605329"
	var password = "Aa123456"

	// 打开登录页面
	err := chromedp.Run(ctx,
		chromedp.Navigate("https://mp.toutiao.com/"),
		chromedp.WaitVisible(`li[aria-label="账密登录"]`),
		chromedp.Click(`li[aria-label="账密登录"]`),
		chromedp.WaitVisible(`input[name="normal-input"]`),
		chromedp.SendKeys(`input[name="normal-input"]`, username),
		chromedp.SendKeys(`input[name="button-input"]`, password),
		chromedp.Click(`span[class="web-login-confirm-info__checkbox"]`),
		chromedp.Click(`button[class="web-login-button"]`),
		// chromedp.WaitVisible(`div[class="mp-menu-wrapper f-min-scroll f-hover-scroll"]`),
	)

	if err != nil {
		return err
	}

	// 等待登陆成功后跳转
	err = WaitForURLContains(ctx, "profile_v4/index", 1500*time.Second)
	return err
}

func publishDrafts(ctx context.Context, controller PublisherController) error {
	var draftsXPath = "//div[contains(@class, 'draft-item')]"
	var editButtonXPath = ".//a[contains(@class, 'op-button') and text()='编辑']"
	var publishButtonXPath = "//button[contains(@class, 'publish-content')]"

	for {
		// 导航到草稿页面
		err := chromedp.Run(ctx,
			chromedp.Navigate("https://mp.toutiao.com/profile_v4/manage/draft"),
			chromedp.WaitVisible("//div[contains(@class, 'list')]"),
		)
		if err != nil {
			return err
		}

		var draftNodes []*cdp.Node
		err = chromedp.Run(ctx,
			chromedp.Nodes(draftsXPath, &draftNodes, chromedp.AtLeast(1)),
		)
		if err != nil || len(draftNodes) == 0 {
			log.Println("没有更多草稿了，等待 1 小时后再重试")
			time.Sleep(1 * time.Hour)
			continue
		}

		for _, draft := range draftNodes {
			_ = draft
			// 等待发布条件
			for {
				canPub, wait := controller.CanPublish(ctx)
				if !canPub {
					time.Sleep(wait)
					continue
				}
				break
			}

			log.Println("处理草稿...")

			// 监听新窗口
			newTargetChan := chromedp.WaitNewTarget(ctx, func(info *target.Info) bool {
				return true
			})

			err = chromedp.Run(ctx,
				chromedp.Click(editButtonXPath),
			)
			if err != nil {
				log.Println("点击编辑按钮失败:", err)
				continue
			}

			// 获取新窗口 ID
			newTarget := <-newTargetChan
			if newTarget == "" {
				log.Println("未能检测到新窗口")
				continue
			}

			// 切换到新标签页
			newTargetCtx, cancel := chromedp.NewContext(ctx, chromedp.WithTargetID(newTarget))
			defer cancel()

			// 执行发布
			err = chromedp.Run(newTargetCtx,
				chromedp.WaitVisible(publishButtonXPath),
				chromedp.MouseClickXY(200, 200), // 收起发文助手
				chromedp.ScrollIntoView(publishButtonXPath),
				chromedp.WaitEnabled(publishButtonXPath),
				chromedp.Click(publishButtonXPath),
				chromedp.Sleep(3*time.Second),
			)
			if err != nil {
				log.Println("发布草稿失败:", err)
				continue
			}

			log.Println("发布成功")

			// 关闭新标签页
			chromedp.Cancel(newTargetCtx)

			// 刷新当前标签页
			err = chromedp.Run(ctx,
				chromedp.Reload(),
			)
			if err != nil {
				log.Println("刷新页面失败:", err)
			}

		}
	}
}

func main() {
	// 创建无头浏览器上下文
	ctx, cancel := createHeadlessDriver()
	defer cancel()

	// 创建发布频率控制器
	// 规则：每小时最多发布8条，间隔3～8分钟
	controller := NewPublisherController(8, 3*time.Minute, 15*time.Minute, &defaultTimeProvider{}, &defaultRandProvider{}, log.Default())

	// 执行登录操作
	if err := waitLogin(ctx); err != nil {
		log.Fatal("登录失败:", err)
	}

	// 发布草稿
	for {
		if err := publishDrafts(ctx, controller); err != nil {
			log.Default().Println("发布草稿失败:", err)
		}

		time.Sleep(1 * time.Hour)
	}

	// log.Println("脚本执行完成")
}

// WaitForURLContains 等待当前页面 URL 包含指定的字符串
func WaitForURLContains(ctx context.Context, substr string, timeout time.Duration) error {
	deadlineCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	for {
		var currentURL string
		err := chromedp.Run(deadlineCtx,
			chromedp.Location(&currentURL),
		)
		if err != nil {
			return err
		}

		if strings.Contains(currentURL, substr) {
			return nil
		}

		time.Sleep(500 * time.Millisecond)
	}
}
